docs/design/ocs-to-ocs-api-server.md

```diff
@@ -0,0 +1,201 @@
+# OCS to OCSs : API Server
+
+## Introduction
+This document defines the API server used for communication between OCS provider and OCS consumer clusters.
+
+
+## Tool
+Following tools can be used to create the API server:
+
+- REST (JSON over HTTP)
+- gRPC
+
+## Tool Comparison
+### gRPC vs REST
+
+- #### Performance:
+    - gRPC is more performant than REST due to binary payloads.
+    - gRPC comes with HTTP2 by default. This has additional benefits like smaller packet size due to compressed headers, multiplexing (same TCP connection for multiple requests), server push streams, etc.
+    - Current use case does not involve frequent requests between client and server. So performance aspect of the gRPC can be ignored.
+
+- #### Security:
+    - Both provide security with TLS support.
+    - gRPC endpoint can't be invoked via browser/postman.
+
+We will use gRPC as we are on k8s and gRPC has more advantages compared to REST
+
+
+## RPC Interface
+
+#### OCSProvider service
+
+``` protobuf
+syntax = "proto3";
+package provider;
+
+import "google/protobuf/descriptor.proto";
+import "google/protobuf/struct.proto";
+
+option go_package = "./;providerpb";
+
+
+// OCSProvider holds the RPC methods that the OCS consumer can use to communicate with remote OCS provider cluster
+service OCSProvider {
+  // OnBoardConsumer RPC call to validate the consumer and create StorageConsumer
+  // resource on the storage provider cluster
+  rpc OnBoardConsumer (OnBoardConsumerRequest)
+  returns (OnBoardConsumerResponse) {}
+  // GetStorageConfig RPC call to generate the json config for connecting to storage provider cluster
+  rpc GetStorageConfig(StorageConfigRequest)
+  returns (StorageConfigResponse) {}
+  // UpdateCapacity RPC call to increase or decrease the storage pool size
+  rpc UpdateCapacity(UpdateCapacityRequest)
+  returns (UpdateCapacityResponse) {}
+  // OffboardConsumer RPC call to delete StorageConsumer CR on the storage provider cluster.
+  rpc OffboardConsumer (OffboardConsumerRequest)
+  returns (OffboardConsumerResponse) {}
+}
+```
+
+#### OnboardConsumer
+- Check if any existing `StorageConsumer` CR with same token is available and return its encrypted UUID
+- If no existing `StorageConsumer` CR found:
+    - Create `StorageConsumer` CR after validating the token.
+    - Return the encrypted UUID of the `StorageConsumer` CR
+
+```protobuf
+// OnBoardConsumerRequest holds the required information to validate the consumer and create StorageConsumer
+// resource on the StorageProvider cluster
+message OnBoardConsumerRequest{
+    // token provided by the provider cluster admin to authenticate the consumer
+    string token =1;
+    // capacity is a valid k8s resource quantity
+    string capacity = 2;
+}
+
+// OnBoardConsumerResponse holds the response for OnBoardConsumer API request
+message OnBoardConsumerResponse{
+    // K8s UID (UUID) of the consumer cluster
+    string storageConsumerUUID =1;
+}
```

Comment:
Do we want to return the granted capacity?

Reply:
adding `grantedCapacity` to `OnBoardConsumerResponse` and `UpdateCapacityResponse`.